salt/modules/boto_s3.py

```diff
@@ -0,0 +1,176 @@
+
+        http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/
+            iam-roles-for-amazon-ec2.html
+
+        http://boto3.readthedocs.io/en/latest/guide/
+            configuration.html#guide-configuration
+
+    If IAM roles are not used you need to specify them either in a pillar or
+    in the minion's config file:
+
+    .. code-block:: yaml
+
+        s3.keyid: GKTADJGHEIQSXMKKRBJ08H
+        s3.key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+
+    A region may also be specified in the configuration:
+
+    .. code-block:: yaml
+
+        s3.region: us-east-1
+
+    If a region is not specified, the default is us-east-1.
+
+    It's also possible to specify key, keyid and region via a profile, either
+    as a passed in dict, or as a string to pull from pillars or minion config:
+
+    .. code-block:: yaml
+
+        myprofile:
+            keyid: GKTADJGHEIQSXMKKRBJ08H
+            key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+            region: us-east-1
+
+:depends: boto3
+'''
+# keep lint from choking on _get_conn and _cache_id
+# pylint: disable=E0602
+
+# Import Python libs
+from __future__ import absolute_import
+import logging
+
+# Import Salt libs
+from salt.utils.versions import LooseVersion as _LooseVersion
+
+log = logging.getLogger(__name__)
+
+# pylint: disable=import-error
+try:
+    # pylint: disable=unused-import
+    import boto3
+    # pylint: enable=unused-import
+    import botocore
+    logging.getLogger('boto3').setLevel(logging.CRITICAL)
+    HAS_BOTO = True
+except ImportError:
+    HAS_BOTO = False
+# pylint: enable=import-error
+
+
+def __virtual__():
+    '''
+    Only load if boto libraries exist and if boto libraries are greater than
+    a given version.
+    '''
+    needed_boto3_version = '1.2.1'
+    msg = 'The boto_s3 module cannot be loaded: {0}.'
+    if not HAS_BOTO:
+        return (False, msg.format('boto3 libraries not found'))
+    if _LooseVersion(boto3.__version__) < _LooseVersion(needed_boto3_version):
+        submsg = 'boto3 library version {0} is required'.format(
+            needed_boto3_version,
+        )
+        return (False, msg.format(submsg))
+    return True
+
+
+def __init__(opts):  # pylint: disable=unused-argument
+    if HAS_BOTO:
+        __utils__['boto3.assign_funcs'](__name__, 's3')
+
+
+def get_object_metadata(
+    name,
+    extra_args=None,
+    region=None,
+    key=None,
+    keyid=None,
+    profile=None,
+):
+    '''
+    Get metadata about an S3 object.
+    Returns None if the object does not exist.
+
+    You can pass AWS SSE-C related args and/or RequestPayer in extra_args.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt myminion boto_s3.get_object_metadata \\
+                         my_bucket/path/to/object \\
+                         region=us-east-1 \\
+                         key=key \\
+                         keyid=keyid \\
+                         profile=profile \\
+    '''
+    bucket, _, key = name.partition('/')
```

Comment:
Is it intended that this `key` var overwrite any value for `key` given in the function params above?  I don't imagine anyone is using objects named after their AWS access keys :)

Can I suggest maybe something like the following to help simplify the code, and avoid the name clash? 
```
def get_object_metadata(
    name,
    region=None,
    key=None,
    keyid=None,
    profile=None,
    **kwargs
):
    '''
    Get metadata about an S3 object.
    Returns None if the object does not exist.

    CLI Example:

    .. code-block:: bash

        salt myminion boto_s3.get_object_metadata \\
                         my_bucket/path/to/object \\
                         region=us-east-1 \\
                         key=key \\
                         keyid=keyid \\
                         profile=profile \\
    '''
    if 'Bucket' not in kwargs:
        kwargs['Bucket'], _, kwargs['Key'] = name.partition('/')

    conn = _get_conn(region=region, key=key, keyid=keyid, profile=profile)

    try:
        metadata = conn.head_object(**kwargs)
    except botocore.exceptions.ClientError as e:
        if e.response['Error']['Message'] == 'Not Found':
            log.debug('S3 object not found: {0}/{1}'.format(kwargs['Bucket'],
                                                            kwargs['Key']))
            return {'result': None}
        return {'error': __utils__['boto3.get_error'](e)}

    return {'result': metadata}
```

Reply:
Ah, this is a mistake I didn't catch while refactoring, I meant to name it `s3_key`. Thanks for catching this! I'll fix it.